tools/OpenGL-Header-Generator/GeneratorGL.py

```diff
@@ -0,0 +1,615 @@
+
+            self.output.AddCommand(info.GetCommand())
+            self.output.AddProc(info.GetProc())
+
+    def _GenerateEnums(self, registry: etree.Element):
+
+        for element in registry.findall("enums/enum"):
+
+            if element.get("name") == None:
+                element.attrib["name"] = element.find("name").text
+
+            info = self.EnumInfo(element)
+
+            if not self.input.ContainsEnum(info.GetName()):
+                continue
+
+            self.output.AddEnum(info.GetOutput())
+
+    def _GenerateTypes(self, registry: etree.Element):
+
+        for element in registry.findall("types/type"):
+
+            if element.get("name") == None:
+                element.attrib["name"] = element.find("name").text
+
+            info = self.TypeInfo(element)
+
+            if info.GetValue() == "":
+                continue
+
+            if info.GetValue().startswith("union"):
+                continue
+
+            if info.GetName().startswith("struct"):
+                continue
+
+            if info.GetValue().startswith("#include"):
+                continue
+
+            if info.GetValue().startswith("#ifdef"):
+                continue
+
+            if info.GetValue().startswith("#ifndef"):
+                continue
+
+            self.output.AddType(info.GetOutput())
+
+    def Process(self):
+
+        tree = etree.parse(self.xml)
+        registry = tree.getroot()
+
+        # Find required typs/enums/commands from features and extensions
+        self._GetFeatures(registry)
+        self._GetExtensions(registry)
+
+        # Generate types/enums/commands from requirements
+        self._GenerateTypes(registry)
+        self._GenerateEnums(registry)
+        self._GenerateCommands(registry)
+
+        template_args = {
+            "platform_include": Generator.template_platform_include,
+            "types": self.output.GetTypeString(),
+            "enums": self.output.GetEnumString(),
+            "procs": self.output.GetProcString(),
+            "methods": self.output.GetCommandString(),
+        }
+
+        # Generate header from generated types/enums/commands
+        self.output_h = Generator.template_h.format(**template_args)
+
+    def Write(self):
+
+        with open(f"{self.output_file}", "w") as f:
+            f.write(self.output_h)
+
+
+class GeneratorGL(Generator):
+    def __init__(self, args) -> None:
+        super().__init__(args)
+
+        self.xml = args.gl_h_xml
+        self.apple = args.gl_apple
+
+        self.apis = {
+            "gl": [
+                "GL_VERSION_1_0",
+                "GL_VERSION_1_1",
+                "GL_VERSION_1_2",
+                "GL_VERSION_1_3",
+                "GL_VERSION_1_4",
+                "GL_VERSION_1_5",
+                "GL_VERSION_2_0",
+                "GL_VERSION_2_1",
+                "GL_VERSION_3_0",
+                "GL_VERSION_3_1",
+            ],
+            "gles2": [
+                "GL_ES_VERSION_2_0",
```

Comment:
Makes it sense to add also `"GL_VERSION_4_0"` for future?
And seems a `GL_ES_VERSION_3_0` becomes needed as Kodi have the support about?

Reply:
I've only added what Kodi currently needs. If someone needs to support a feature from `GL_VERSION_4_0` then it can be added :) The same thing goes for `GL_ES_VERSION_3_0`.

Reply:
Kodi never used GLES 3.0? OK, was then wrong by me :smile:.

Was come to it as e.g. this need it https://github.com/xbmc/screensavers.rsxs/blob/Nexus/CMakeLists.txt#L73-L81

And in Kodi itself also checked about 3.0 here https://github.com/xbmc/xbmc/blob/master/cmake/modules/FindOpenGLES.cmake#L30-L37.

Reply:
I originally added gles3 for vaapi with the gl sync code. I have replaced that with egl fences so gles3 isn't required. There are a couple other reference to gles3 in the codebase but they can be removed.

The other problem is that there is build time and runtime requirements. We only ask for a gles 2.0 context when running so that should probably be checked if needing gles3 features.